### PR TITLE
fix(core): preserve trailing newlines in shell execution

### LINKED ISSUE #20755: Shell Tool Trailing Newline Trimming
## Description
   The run_shell_command tool erroneously trims trailing newlines from command
   strings before execution. This results in syntax errors for shell constructs
   that require a trailing newline as a delimiter, most notably Bash heredocs.

   Bash expects heredoc  delimiters to be followed by a newline to signal the end of the input block.
   Without it, the shell encounters an unexpected EOF while still looking for  the delimiter.

 ## Examples

   ### Heredoc Syntax Error (Triggered by Newline Trimming)
   **Command Sent:**
```console
$  cat <<EOF
test
EOF
 bash: warning: here-document delimited by end-of-file (wanted `EOF')
 bash: syntax error: unexpected end of file
```
   ### Heredoc Workaround (The ":" Fix)

   Adding a command on a subsequent line preserves the newline after the heredoc
   delimiter.
``` console
$  cat <<EOF
test
EOF
:
test
```

   ### Non-Heredoc Syntax Error (Single Comment)
   A command string containing only a comment fails when the newline is trimmed
   because the shell identifies the input as empty or truncated in certain
   execution contexts.
```console
$  # comment
bash: syntax error: unexpected end of file
```
[ACTION REQUIRED] 📎 PLEASE ATTACH THE EXPORTED CHAT HISTORY JSON FILE TO THIS ISSUE IF YOU FEEL COMFORTABLE SHARING IT.

### What did you expect to happen?

run_shell_tool should not trim trailing newlines

### Client information

* **CLI Version:** 0.31.0
* **Git Commit:** 72c653354
* **Session ID:** b119e6e0-a35b-488a-8978-38d2df50c7ec
* **Operating System:** linux v25.6.0
* **Sandbox Environment:** no sandbox
* **Model Version:** auto-gemini-3
* **Auth Type:** oauth-personal
* **Memory Usage:** 533.8 MB
* **Terminal Name:** Unknown
* **Terminal Background:** #002b36
* **Kitty Keyboard Protocol:** Unsupported

### Login information

Oauth

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by Aaxhirrr**
## Summary

Fixes a regression in `run_shell_command` where trailing newlines were stripped before execution, which breaks heredocs and other multiline shell constructs that require a final newline delimiter.

## Details

This change keeps the fix narrowly scoped to the execution path.

- Preserve trailing whitespace only when preparing the command for execution.
- Keep existing trimmed behavior for confirmation text, command-root parsing, and allowlist/policy-related parsing so the patch has a smaller blast radius.
- Update the Unix `pgrep` wrapper so newline-terminated commands close cleanly instead of being trimmed and forced onto a same-line `;`.
- Add focused regression coverage for:
  - wrapped heredoc commands
  - wrapped comment-only commands
  - Linux shell execution wrapping
  - a bash-only integration case driven by fake responses

No documentation changes were needed.

## Related Issues

Fixes #20755

## How to Validate

1. Run the core shell regressions:

   ```bash
   npm run test --workspace @google/gemini-cli-core -- shell.test.ts shell-utils.test.ts
   ```

   Expected result:
   - `shell.test.ts` passes
   - `shell-utils.test.ts` passes

2. In a bash environment, run the targeted integration regression:

   ```bash
   npm run test:integration:sandbox:none -- run_shell_command.test.ts -t "should preserve trailing newlines for heredoc shell commands"
   ```

   Expected result:
   - the test passes
   - the recorded `run_shell_command` args still end with `\n`
   - the output contains `TRAILING_NEWLINE_20755`
   - the output does not contain `here-document delimited by end-of-file` or `syntax error: unexpected end of file`

3. Manual sanity check in a bash environment:

   Execute a shell tool call with:

   ```bash
   cat <<EOF
   test
   EOF
   ```

   Expected result:
   - command succeeds
   - output contains `test`
   - no heredoc EOF or syntax error is emitted

4. Optional extra edge case:

   Execute a comment-only command ending in a newline:

   ```bash
   # comment
   ```

   Expected result:
   - no unexpected EOF caused by trimming

Note:
- The new integration regression is bash-only, so it is skipped on Windows.
- I validated the unit test coverage locally on Windows with `npm run`.

## Pre-Merge Checklist

- [ ] Updated relevant documentation and README (if needed)
- [x] Added/updated tests (if needed)
- [ ] Noted breaking changes (if any)
- [ ] Validated on required platforms/methods:
  - [ ] MacOS
    - [ ] npm run
    - [ ] npx
    - [ ] Docker
    - [ ] Podman
    - [ ] Seatbelt
  - [x] Windows
    - [x] npm run
    - [ ] npx
    - [ ] Docker
  - [ ] Linux
    - [ ] npm run
    - [ ] npx
    - [ ] Docker

### CODE REVIEW

File: packages/core/src/tools/shell.ts [hunk 1]
```diff
@@ -182,9 +184,13 @@ export class ShellToolInvocation extends BaseToolInvocation<
...
-            let command = strippedCommand.trim();
-            if (!command.endsWith('&')) command += ';';
-            return `{ ${command} }; __code=$?; pgrep -g 0 >${tempFilePath} 2>&1; exit $__code;`;
+            let command = strippedCommand;
+            const hasTrailingLineTerminator = /[\r\n]$/.test(command);
+            if (!command.endsWith('&') && !hasTrailingLineTerminator) {
+              command += ';';
+            }
+            const closingGroup = hasTrailingLineTerminator ? '}' : ' }';
+            return `{ ${command}${closingGroup}; __code=$?; pgrep -g 0 >${tempFilePath} 2>&1; exit $__code;`;
```
[the remark this answers is not in the document]
Comment (Aaxhirrr) [reply packages/core/src/tools/shell.ts]: Resolved.

File: packages/core/src/tools/shell.ts [hunk 2]
```diff
@@ -236,9 +239,15 @@ export class ShellToolInvocation extends BaseToolInvocation<
         ? strippedCommand
         : (() => {
             // wrap command to append subprocess pids (via pgrep) to temporary file
-            let command = strippedCommand.trim();
-            if (!command.endsWith('&')) command += ';';
-            return `{ ${command} }; __code=$?; pgrep -g 0 >${tempFilePath} 2>&1; exit $__code;`;
+            let command = strippedCommand;
+            const hasTrailingLineTerminator = /[\r\n]$/.test(command);
```
Comment (mrpmohiburrahman) [inline packages/core/src/tools/shell.ts]: Here, this matches either `\r` or `\n` as the last character. Since this code path
only runs on non-Windows `(line 238)`, the only line terminator bash actually
understands is `\n`. A bare `\r` is just a regular character to bash - it
doesn't separate tokens.

So if a command were to end with a bare `\r` (no `\n`), the wrapping would
produce:

```
 { echo hello\r}; __code=$?; ...
```

Bash would read `\r}` as one word instead of seeing `}` as the group-closing
keyword, leaving the `{` unmatched and causing a syntax error.

A small tweak would prevent this:

```typescript
const hasTrailingLineTerminator = /\r?\n$/.test(command);
```

This still handles `\n` (Unix) and `\r\n` (Windows-style line endings pasted
into a Unix session), but a bare `\r` would correctly fall through to the
regular `command += ';'` path where it's harmless.

Comment (Aaxhirrr) [reply packages/core/src/tools/shell.ts]: Thanks, this was a good catch.

I updated the wrapper logic so it only treats `\n` or `\r\n` as a real trailing line terminator, and a bare `\r` now goes through the normal path instead of being treated like a newline. I also added a regression test for that case while keeping the heredoc and comment-only trailing-newline coverage.

I rebased the branch and pushed the updated fix. Could you please take another look when you get a chance?

**Comment by scidomino** [comment]
Closing as this seems  to have been fixed in https://github.com/google-gemini/gemini-cli/pull/25537
